src/scenes/pt.cu

```diff
@@ -183,29 +225,84 @@ CU_HD Intersection Disc::intersect(Ray const &r, Object const *obj) const {
   return {t, obj, x, normal};
 }
 
+CU_D ObjectSample Object::sample(Vec3 const &dir, TraceContext &ctx) const {
+  ObjectSample sample;
+  if (type == SPHERE) {
+    sample = sphere.sample(dir, ctx, this);
+  } else if (type == PLANE) {
+    sample = plane.sample(dir, ctx, this);
+  } else if (type == DISC) {
+    sample = disc.sample(dir, ctx, this);
+  }
+  return sample;
+}
+
+CU_D ObjectSample Sphere::sample(Vec3 const &hitx, TraceContext &ctx,
+                                 [[maybe_unused]] Object const *obj) const {
+  Vec3 p = uniformHemisphereSampling(ctx);
+  p      = onb(p, obj->tr.translation() - hitx);
+  Vec3 n = p.normalized();
+  p      = radius * p + obj->tr.translation();
+
+  float lightAngle = atan(radius / (p - hitx).norm());
+  float solidAngle = R_PI * lightAngle * lightAngle;
+
+  return {1 / solidAngle, p, n};
+}
+CU_D ObjectSample Plane::sample(Vec3 const &dir, TraceContext &ctx,
+                                [[maybe_unused]] Object const *obj) const {
+  // TODO
+  return {};
+}
+CU_D ObjectSample Disc::sample(Vec3 const &dir, TraceContext &ctx,
+                               [[maybe_unused]] Object const *obj) const {
+  // TODO
+  return {};
+}
 /**********************************
  * BSDF
  **********************************/
+CU_D Vec3 uniformHemisphereSampling(TraceContext &ctx) {
+  float r     = sqrt(ctx.sample1D());
+  float theta = ctx.sample1D() * 2 * R_PI;
+  Vec3 p      = {r * cos(theta), 0, r * sin(theta)};
+  p           = {p.x(), sqrt(fmaxf(0.0f, 1.0 - p.x() * p.x() - p.z() * p.z())), p.z()};
+  return p;
+}
+CU_D Vec3 onb(Vec3 const &o, Vec3 const &dir) {
```

Comment:
Might be better to return a struct with the 3 vectors and then multiply with that separately. Since in other code you might need to use the same basis multiple times.